flink-connector-opensearch/src/main/java/org/apache/flink/connector/opensearch/sink/DocSerdeRequest.java

```diff
@@ -0,0 +1,107 @@
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.connector.opensearch.sink;
+
+import org.apache.flink.annotation.PublicEvolving;
+
+import org.opensearch.action.DocWriteRequest;
+import org.opensearch.action.delete.DeleteRequest;
+import org.opensearch.action.index.IndexRequest;
+import org.opensearch.action.update.UpdateRequest;
+import org.opensearch.common.bytes.BytesReference;
+import org.opensearch.common.io.stream.BytesStreamOutput;
+import org.opensearch.common.io.stream.InputStreamStreamInput;
+import org.opensearch.common.io.stream.StreamInput;
+import org.opensearch.common.io.stream.StreamOutput;
+
+import java.io.DataInputStream;
+import java.io.DataOutputStream;
+import java.io.IOException;
+import java.io.Serializable;
+
+/**
+ * Wrapper class around {@link DocWriteRequest} since it does not implement {@link Serializable},
+ * required by AsyncSink scaffolding.
+ *
+ * @param <T> type of the write request
+ */
+@PublicEvolving
+public class DocSerdeRequest<T> implements Serializable {
+    private static final long serialVersionUID = 1L;
+    private final DocWriteRequest<T> request;
+
+    private DocSerdeRequest(DocWriteRequest<T> request) {
+        this.request = request;
+    }
+
+    public DocWriteRequest<T> getRequest() {
+        return request;
+    }
+
+    static <T> DocSerdeRequest<T> from(DocWriteRequest<T> request) {
+        return new DocSerdeRequest<>(request);
+    }
+
+    static DocSerdeRequest<?> readFrom(long requestSize, DataInputStream in) throws IOException {
+        try (final StreamInput stream = new InputStreamStreamInput(in, requestSize)) {
+            return new DocSerdeRequest<>(readDocumentRequest(stream));
+        }
+    }
+
+    void writeTo(DataOutputStream out) throws IOException {
+        try (BytesStreamOutput stream = new BytesStreamOutput()) {
+            writeDocumentRequest(stream, request);
+            out.write(BytesReference.toBytes(stream.bytes()));
+        }
+    }
+
+    /** Read a document write (index/delete/update) request. */
+    private static DocWriteRequest<?> readDocumentRequest(StreamInput in) throws IOException {
+        byte type = in.readByte();
+        DocWriteRequest<?> docWriteRequest;
+        if (type == 0) {
+            docWriteRequest = new IndexRequest(in);
+        } else if (type == 1) {
+            docWriteRequest = new DeleteRequest(in);
+        } else if (type == 2) {
+            docWriteRequest = new UpdateRequest(in);
+        } else {
+            throw new IllegalStateException("Invalid request type [" + type + " ]");
+        }
+        return docWriteRequest;
```

Comment:
These methods are untested. Should we add unit tests for them?

Reply:
They are tested in scope if integration test, OpensearchAsyncSinkITCase, both reading and writing side.

Reply:
+1 for a unit test. Unless there is a good reason not to, unit tests give quicker feedback.